### PR TITLE
ROCm: Improve compiler

### PR DESCRIPTION
**Pull request by leofang**
- Avoid using `hipcc` for preprocessing if possible.
- Use hipRTC's version as part of cache key (as is done with NVRTC)
- Fix header path search in ROCm
- Reduce amount of "hipify" work that we do with regard to kernel syntax

### CODE REVIEW

File: cupy/cuda/compiler.py [hunk 1]
```diff
@@ -703,18 +694,30 @@ def _preprocess_hipcc(source, options):
...
 
 def _convert_to_hip_source(source, extra_source, is_hiprtc):
-    table = [
-        ('threadIdx.', 'hipThreadIdx_'),
-        ('blockIdx.', 'hipBlockIdx_'),
-        ('blockDim.', 'hipBlockDim_'),
-        ('gridDim.', 'hipGridDim_'),
-    ]
-    for i, j in table:
-        source = source.replace(i, j)
```
Comment (leofang) [inline cupy/cuda/compiler.py:717]: In HIP's official test suite, they've even abandoned their own quirky syntax `hipThreadIdx_x` in favor of CUDA's...

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
@emcastillo I think this is ready. The PR description is updated. PTAL.

**Comment by leofang** [comment]
Conflict resolved.

Jenkins, test this please

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1952,36 +1956,48 @@ cpdef function.Module compile_with_cache(
...
+        else:
+            # CUDA versions not yet supported.
+            bundled_include = None
 
-    if bundled_include is not None:
-        options += ('-I' + os.path.join(
-            _get_header_dir_path(), 'cupy', '_cuda', bundled_include),)
+        if bundled_include is None and _cuda_path is None:
+            raise RuntimeError(
+                'Failed to auto-detect CUDA root directory. '
```
Comment (emcastillo) [inline cupy/core/core.pyx:1985]: This message is CUDA specific and can lead to confusion to HIP users :)

Comment (leofang) [reply cupy/core/core.pyx:1985]: This branch is CUDA-only 😁 But let me fix the error message, as I messed up during merging...

Comment (leofang) [reply cupy/core/core.pyx:1985]: Done in 3e48758.

**Comment by leofang** [comment]
Jenkins, test this please

**Review by emcastillo — APPROVED** [review]

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**